examples/src/main/java/com/mapbox/navigation/examples/core/MapMatchingActivity.kt

```diff
@@ -0,0 +1,396 @@
+        directionsRoute = getRouteFromBundle(savedInstanceState)
+    }
+
+    fun startLocationUpdates() {
+        val request = LocationEngineRequest.Builder(INTERVAL_MILLISECONDS)
+            .setFastestInterval(FASTEST_INTERVAL_MILLISECONDS)
+            .setPriority(LocationEngineRequest.PRIORITY_HIGH_ACCURACY)
+            .build()
+
+        localLocationEngine.requestLocationUpdates(
+            request,
+            locationEngineCallback,
+            Looper.getMainLooper()
+        )
+
+        if (directionsRoute == null) {
+            localLocationEngine.getLastLocation(locationEngineCallback)
+        }
+    }
+
+    private fun stopLocationUpdates() {
+        localLocationEngine.removeLocationUpdates(locationEngineCallback)
+    }
+
+    private val tripSessionStateObserver = object : TripSessionStateObserver {
+        override fun onSessionStateChanged(tripSessionState: TripSessionState) {
+            when (tripSessionState) {
+                STARTED -> stopLocationUpdates()
+                STOPPED -> {
+                    startLocationUpdates()
+                    navigationMapboxMap?.removeRoute()
+                    updateCameraOnNavigationStateChange(false)
+                }
+            }
+        }
+    }
+
+    private fun updateCameraOnNavigationStateChange(navigationStarted: Boolean) {
+        navigationMapboxMap?.apply {
+            if (navigationStarted) {
+                updateCameraTrackingMode(NavigationCamera.NAVIGATION_TRACKING_MODE_GPS)
+                updateLocationLayerRenderMode(RenderMode.GPS)
+            } else {
+                updateCameraTrackingMode(NavigationCamera.NAVIGATION_TRACKING_MODE_NONE)
+                updateLocationLayerRenderMode(RenderMode.COMPASS)
+            }
+        }
+    }
+
+    private inner class MyLocationEngineCallback(activity: MapMatchingActivity) :
+        LocationEngineCallback<LocationEngineResult> {
+
+        private val activityRef = WeakReference(activity)
+
+        override fun onSuccess(result: LocationEngineResult) {
+            activityRef.get()?.locationComponent?.forceLocationUpdate(result.lastLocation)
+        }
+
+        override fun onFailure(exception: java.lang.Exception) {
+            Timber.i(exception)
+        }
+    }
+
+    private inner class TestUrlSkuTokenProvider : UrlSkuTokenProvider {
+        override fun obtainUrlWithSkuToken(resourceUrl: String, querySize: Int): String {
+            val urlBuilder = StringBuilder(resourceUrl)
+            when (querySize == 0) {
+                true -> urlBuilder.append("?")
+                false -> urlBuilder.append("&")
+            }
+            urlBuilder.append("sku=${Utils.getMapboxAccessToken(applicationContext)}")
+            return urlBuilder.toString()
+        }
+    }
+
+    private inner class TestRerouteController : RerouteController {
+        override val state = RerouteState.Idle
+
+        override fun reroute(routesCallback: RerouteController.RoutesCallback) {
```

Comment:
Looking at this wondering if `RerouteController` exposes a friendly API 🤔 

`RerouteActivity` relies on `routesCallback` and calls to `onNewRoutes` which underneath calls `MapboxNavigation#setRoutes` 👀 https://github.com/mapbox/mapbox-navigation-android/blob/81903d84497b9aee88321625e39d339b80e86f78/examples/src/main/java/com/mapbox/navigation/examples/core/ReRouteActivity.kt#L376 That's another option that would remove the need of calling `setRoutes` explicitly here https://github.com/mapbox/mapbox-navigation-android/blob/67241fc3a4a21da2ec6d9c7f3c90d9d477eb1929/examples/src/main/java/com/mapbox/navigation/examples/core/MapMatchingActivity.kt#L218 and also would use the callback given. Not sure though if its behavior is clear from a client point of view. We should probably specify in the docs. Wondering even if we should provide an API without the `RerouteController.RoutesCallback` and let clients call `setRoutes` explicitly 🤔 cc @RingerJK

Reply:
Interface `RoutesCallback` with docs:

```
     /**
     * Route Callback is useful to set new route(s) on reroute event. Doing the same as
     * [MapboxNavigation.setRoutes].
     */
    interface RoutesCallback {
        /**
         * Called whenever new route(s) has been fetched.
         * @see [MapboxNavigation.setRoutes]
         */
        fun onNewRoutes(routes: List<DirectionsRoute>)
    }
```

So here we mentioned that `onNewRoutes` actually the same as `MapboxNavigation.setRoutes`.

Reply:
to fetch a route we have the next flow:
get destination -> get route with `DirectionsApi` -> take `maneuvers` coordinates -> get route with `MapMatching` -> `navigation.setRoutes`
we use the same flow on `reroute` event. So, in that case no need to use `RerouteController.RoutesCallback`, because we set routes directly

Reply:
`RoutesCallback` is support for customers to not provide own callback and/or not hold `MapboxNavigation` instance at own RerouteController. It's up 2 u how to provide a new route(via `RoutesCallback` or `MapboxNavigation.setRoute`)

Reply:
This is exactly what I'm questioning, isn't having two different ways of doing "the same" confusing? Is `RoutesCallback` here worth it? Should the `RerouteController` APIs have a different design? 🤔

Reply:
Should we cut a separate ticket to keep track of this discussion? It doesn't seem that fits in https://github.com/mapbox/mapbox-navigation-android/issues/3450